### PR TITLE
Implement MySQL Frontend based on Netty

### LINKED ISSUE #1219: [Umbrella] MySQL Protocol Frontend
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues) and found no similar issues.

### Describe the proposal

Implement MySQL Protocol Frontend service in `kyuubi-server` to support connect Kyuubi by MySQL client, e.g. `mysql` CLI, MySQL JDBC driver.

image

### Task list

- [ ] #1348
- [ ] #1358
- [ ] #1395
- [ ] #1376

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

Close #1219.

Implment MySQL text protocol, support basic query use mysql cli to run Spark SQL.

Test:

Add `MySQLSparkQuerySuite`.

Change `OutputSchemaTPCDSSuite` from **Thrift binary protocol** to **MySQL protocol**.

Limitations:

1) only support mysql_native_password authentication mechanism
2) only support MySQL text protocol
3) not support send back result which row larger than 16M
4) not support server-side prepared statement
5) not support overwrite engine_conf/session_conf via JDBC url parameters like Hive JDBC client
6) limited support MySQL metadata queries, thus some MySQL GUI client which send metadata queries implicitly may not work as expected
7) not support SSL

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala [hunk 1]
```diff
@@ -77,7 +79,11 @@ class KyuubiMySQLFrontendService(override val serverable: Serverable)
       .childHandler(new ChannelInitializer[SocketChannel] {
         override def initChannel(channel: SocketChannel): Unit = channel.pipeline
           .addLast(new LoggingHandler("org.apache.kyuubi.server.mysql.codec", LogLevel.TRACE))
-        // TODO implement authentication, codec, command handler
+          .addLast(new MySQLFrameDelimiter)
+          .addLast(new MySQLPacketEncoder)
+          .addLast(new MySQLAuthHandler)
+          .addLast(new MySQLPacketDecoder)
+          .addLast(new MySQLCommandHandler(serverable.backendService, execPool))
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala:86]: The netty pipeline can be an entrypoint of this PR for review

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/mysql/MySQLCommandHandler.scala [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        val connectionId = ctx.channel.attr(CONNECTION_ID).get
+        connIdToSessHandle.put(connectionId, sessionHandle)
+      }
+    }
+
+  def openSession(ctx: ChannelHandlerContext): SessionHandle = synchronized {
+    try {
+      val user = ctx.channel.attr(USER).get
+      val remoteIp = ctx.channel.attr(REMOTE_IP).get
+      // TODO parse SET command, save other variables at ChannelHandlerContext
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/mysql/MySQLCommandHandler.scala:97]: mysql text protocol does not support pass session confs and vars when open connection

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/mysql/MySQLCommandHandler.scala [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+      val sessionConf = Option(ctx.channel.attr(DATABASE).get) match {
+        case Some(db) => Map("use:database" -> db)
+        case None => Map.empty[String, String]
+      }
+      // v1 is sufficient now, upgrade version when needed
+      val proto = TProtocolVersion.HIVE_CLI_SERVICE_PROTOCOL_V1
+      val sessionHandle = be.openSession(proto, user, "", remoteIp, sessionConf)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/mysql/MySQLCommandHandler.scala:104]: Auth type of BE always be NONE

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/mysql/MySQLDialectHelper.scala [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+package org.apache.kyuubi.server.mysql
+
+import io.netty.channel.ChannelHandlerContext
+
+import org.apache.kyuubi.server.mysql.constant._
+
+object MySQLDialectHelper {
+
+  def convertQuery(origin: String): String = origin.toLowerCase.trim match {
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/mysql/MySQLDialectHelper.scala:26]: A fragile approach to cover parts of metadata queries

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/mysql/authentication/MySQLNativePassword.scala [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+  }
+}
+
+class MySQLNativePassword {
+  final val pluginData = new PluginData
+
+  def login(user: String, host: String, authResp: Array[Byte], database: String):
+  Option[MySQLErrorCode] = {
+    if (isPasswordRight("kyuubi", authResp)) {
+    // if (true) {
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/mysql/authentication/MySQLNativePassword.scala:43]: hardcode now

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
LGTM

**Comment by yaooqinn** [comment]
reopen to check style

**Pull request closed without merging**